[BE][DDP] Fix inplace modification issue

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70260

When DDP forward pass is called twice (i.e. in checkpointing) and module has buffers (i.e. BatchNorm layer), DDP fails with the error:

```
 RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation
```

See https://github.com/pytorch/pytorch/issues/66504, https://discuss.pytorch.org/t/why-i-occur-runtimeerror-if-i-use-distributeddataparallel/139793 for examples where this has occured in the wild.

The problematic line is https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/comm.cpp#L34, where we do an inplace copy_ that autograd raises an error about.

The fix is to avoid the inplace `copy_` and instead directly assign the tensor to the output. This should be functionally the same and the assignment should bump the refcount of the tensor and it shouldn't be freed from underneath us.

Differential Revision: [D33262177](https://our.internmc.facebook.com/intern/diff/D33262177/)

rohan-varma committed:
[BE][DDP] Fix inplace modification issue

When DDP forward pass is called twice (i.e. in checkpointing) and module has buffers (i.e. BatchNorm layer), DDP fails with the error:

```
 RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation
```

See https://github.com/pytorch/pytorch/issues/66504, https://discuss.pytorch.org/t/why-i-occur-runtimeerror-if-i-use-distributeddataparallel/139793 for examples where this has occured in the wild.

The problematic line is https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/comm.cpp#L34, where we do an inplace copy_ that autograd raises an error about.

The fix is to avoid the inplace `copy_` and instead directly assign the tensor to the output. This should be functionally the same and the assignment should bump the refcount of the tensor and it shouldn't be freed from underneath us.

Differential Revision: [D33262177](https://our.internmc.facebook.com/intern/diff/D33262177/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -31,7 +31,7 @@ class BroadcastWork {
     TORCH_INTERNAL_ASSERT(output_tensors.size() == bucket_tensors_.size());
     for(const auto i : c10::irange(output_tensors.size())) {
-      bucket_tensors_[i].copy_(output_tensors[i], /*non_blocking=*/true);
+      bucket_tensors_[i] = output_tensors[i];
```
I originally had one concern around change of semantics here where output_tensors could be modified which would also affect bucket_tensors (which wouldn't happen before this PR). However, after looking closely at how output_tensors are built, I realized this:

1. In the case where we have multiple bucket tensors, flat_tensor_ is actually a brand new Tensor and is only used in this class. `output_tensors` are just views of this flat_tensor_ and hence there is no harm in just assigning like this.
2. In case there is just a single bucket tensor, flat_tensor_ is basically the same as `bucket_tensor[0]` and the assignment is harmless.

Although, this copy was introduced in https://github.com/pytorch/pytorch/pull/20234, so I'll let @mrshenli double check this and stamp it.

mrshenli replied:
hmm, I might miss sth. I was assuming this change would make `bucket_tensors_[i]` point to the new tensor/view `output_tensors[i]` instead of updating the content in the input tensors. So, this might behave differently than before if users have created views from the input tensor (i.e., `bucket_tensors `) before calling this API?

rohan-varma committed:
Update on "[BE][DDP] Fix inplace modification issue"


When DDP forward pass is called twice (i.e. in checkpointing) and module has buffers (i.e. BatchNorm layer), DDP fails with the error:

```
 RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation
```

See https://github.com/pytorch/pytorch/issues/66504, https://discuss.pytorch.org/t/why-i-occur-runtimeerror-if-i-use-distributeddataparallel/139793 for examples where this has occured in the wild.

The problematic line is https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/comm.cpp#L34, where we do an inplace copy_ that autograd raises an error about.

The fix is to avoid the inplace `copy_` and instead directly assign the tensor to the output. This should be functionally the same and the assignment should bump the refcount of the tensor and it shouldn't be freed from underneath us.

Differential Revision: [D33262177](https://our.internmc.facebook.com/intern/diff/D33262177/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[BE][DDP] Fix inplace modification issue"


When DDP forward pass is called twice (i.e. in checkpointing) and module has buffers (i.e. BatchNorm layer), DDP fails with the error:

```
 RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation
```

See https://github.com/pytorch/pytorch/issues/66504, https://discuss.pytorch.org/t/why-i-occur-runtimeerror-if-i-use-distributeddataparallel/139793 for examples where this has occured in the wild.

The problematic line is https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/comm.cpp#L34, where we do an inplace copy_ that autograd raises an error about.

The fix is to avoid the inplace `copy_` and instead directly assign the tensor to the output. This should be functionally the same and the assignment should bump the refcount of the tensor and it shouldn't be freed from underneath us.

Differential Revision: [D33262177](https://our.internmc.facebook.com/intern/diff/D33262177/)

[ghstack-poisoned]
